File: one_dimension/payoffs.py
import numpy as np

"""
Defines several payoff functions.
"""

def make_call_payoff(K):
    """
    Returns a call payoff function with strike K.
    The call payoff works for a scalar value as wel as an np.array of values.
    """
    return lambda x: np.maximum(np.array(x)-K, 0)

def identity_payoff(X_T):
    """
    Identity payoff function. This can be used to reduce the complexity of psi
    in the Exact method for testing purposes.
    
    Args:
        X_T (int or float or list or np.array): The ending value of the process for each path
    Returns:
        (int or float or list or np.array): X_T
    """
    return X_T

File: one_dimension/payoffs.r
# Defines several payoff functions.

make_call_payoff <- function(K)
    # Returns a call payoff function with strike K.
    # The call payoff works for a scalar value as well as an array of values.
    function(x) pmax(x-K, 0)

identity_payoff <- function(X_T)
    # Identity payoff function. This can be used to reduce the complexity of psi
    # in the Exact method for testing purposes.
    #
    # Args:
    #     X_T (numeric or array-like): The ending value of the process for each path
    # Returns:
    #     (int or float or array-like): X_T
    return(X_T)